kfctl/README.md

```diff
@@ -0,0 +1,325 @@
+	http.Handle("/kfctl/iam/apply", optionsHandler(applyIamHandler))
+        
+	http.Handle("/kfctl/initProject", optionsHandler(initProjectHandler))
+	http.Handle("/kfctl/e2eDeploy", optionsHandler(deployHandler))
+```
+
+These functions mostly call a [KsService Interface](https://github.com/kubeflow/kubeflow/blob/master/bootstrap/cmd/bootstrap/app/ksServer.go#L60) to build a ksonnet application.
+The KsService Interface as is cannot be leveraged by kfctl since it includes specific GCP/IAM parameters
+The interface is implemented by [ksServer](https://github.com/kubeflow/kubeflow/blob/master/bootstrap/cmd/bootstrap/app/ksServer.go#L80) which also binds additional methods like [appGenerate](https://github.com/kubeflow/kubeflow/blob/master/bootstrap/cmd/bootstrap/app/ksServer.go#L566) 
+that are relevant to kfctl but cannot be easily separated. It turns out the primary flow
+of interest is createAppHandler. This calls an [anonymous function](https://github.com/kubeflow/kubeflow/blob/master/bootstrap/cmd/bootstrap/app/ksServer.go#L1038) that ends up making all the ksonnet calls required by kfctl.
+The other entrypoints are either not relevent to kfctl or implement part of what is done in 
+this anonymous function.
+
+### Analysis
+
+Methods from ksServer that are relevant to kfctl are:
+- ksServer.CreateApp
+- ksServer.GetApp
+- ksServer.appGenerate
+- ksServer.createComponent
+
+Within these methods there are direct calls to ksonnet 
+- Load
+- RunEnvSet
+- RunInit
+- RunRegistryAdd
+- RunPkgInstall
+- RunPrototypeUse
+- RunParamSet
+- RunApply
+
+The primary type of interest to kfctl is AppConfig (see appendix).
+
+The createAppHandler call sequence is below (only relevant calls are shown):
+
+createAppHandler
+```
+Calls KsService.CreateApp(Context, CreateRequest, Deployment) //not relevant to kfctl
+  Calls KsService.GetProjectLock //not relevant to kfctl
+  Calls KsService.GetApp //not relevant to kfctl
+    Calls kApp.Load //RELEVENT
+    Returns appInfo, repoDir //RELEVENT
+  If ksonnet application exists 
+    Calls kApp.RunEnvSet //`ks env set` RELEVANT
+  Else
+    Calls kApp.RunInit //`ks init <app>` RELEVANT
+  Calls kApp.Load //RELEVANT
+
+  //RELEVANT
+  for CreateRequest.AppConfig.Registries //AppConfig is RELEVANT
+    Calls kApp.RunRegistryAdd //`ks registry add` RELEVANT
+
+  Calls KsService.appGenerate(kApp, AppConfig)  //RELEVANT
+    Calls kApp.Libraries //RELEVANT
+    for AppConfig.Registries
+      Calls kApp.RunPkgInstall //RELEVANT
+    for AppConfig.Packages
+      Calls kApp.RunPkgInstall //RELEVANT
+    for AppConfig.Components //RELEVANT
+      Calls KsService.createComponent //RELEVANT
+
+  Calls KsService.Apply //RELEVANT
+```
+
+- Note: Where there is a RELEVENT comment indicates where we need to insert a shared Interface that kfctl can use.
+- Note: kApp refers to the ksonnet interface.
+
+## Proposed Design
+
+### 1. Move the common interface and types required for the library to a pkg directory under bootstrap 
```

Comment:
Should we define an interface for a Kubeflow App that would wrap ksonnet?

e.g

interface KsApp {
   generateK8s
  generateK8s
   applyK8s
   applyPlatform
}

This struct is what would be used by the CLI and webapp?

I guess an open question is how we would extend it to allow implementations for different platforms?

Reply:
> Should we define an interface for a Kubeflow App that would wrap ksonnet?
> 
> e.g
> 
> interface KsApp {
> generateK8s
> generateK8s
> applyK8s
> applyPlatform
> }
> 
> This struct is what would be used by the CLI and webapp?
> 
> I guess an open question is how we would extend it to allow implementations for different platforms?

Yes I'll define a struct or interface. I'll post here when the implementation is slightly farther along (a day or so hopefully)

Reply:
SGTM. You could also just submit a skeleton PR with just the interface or struct but no implementation first.

Reply:
+1 on getting the skeleton PR in early for review.